File: lc/coreprep/implementation/patterns_reference_implementation/k_way_merge/MergeKSortedArrays.java
import java.util.*;
import java.util.concurrent.*;
import java.util.function.*;

public class MergeKSortedArrays {
  public static void main(String[] args) {
    try (Scanner in = new Scanner(System.in)) {
      int tt = in.nextInt();

      while (tt-- > 0) {
        int n = in.nextInt();

        List<Integer[]> lists = new ArrayList<>();
        for (int i = 0; i < n; i++) {
          int m = in.nextInt();

          Integer[] a = new Integer[m];
          for (int j = 0; j < m; j++) {
            a[j] = in.nextInt();
          }
          lists.add(a);
        }

        var sorted = mergeKSortedArrays(lists);
        for (int s : sorted) {
          System.out.printf("%d ", s);
        }
        System.out.println();
      }
    }
  }

  static class Entry {
    int listIndex;
    int arrayIndex;

    Entry(int listIndex, int arrayIndex) {
      this.listIndex = listIndex;
      this.arrayIndex = arrayIndex;
    }
  }

  // O(nlogk) / O(k)
  public static List<Integer> mergeKSortedArrays(List<Integer[]> lists) {
    PriorityQueue<Entry> minHeap = new PriorityQueue<>(
        (Entry e1, Entry e2)
            -> Integer.compare(lists.get(e1.listIndex)[e1.arrayIndex],
                               lists.get(e2.listIndex)[e2.arrayIndex]));

    int k = lists.size();
    for (int i = 0; i < k; i++) {
      if (lists.get(i) != null) {
        minHeap.add(new Entry(i, 0));
      }
    }

    List<Integer> res = new ArrayList<>();
    while (!minHeap.isEmpty()) {
      Entry entry = minHeap.poll();

      res.add(lists.get(entry.listIndex)[entry.arrayIndex]);
      if (entry.arrayIndex < lists.get(entry.listIndex).length - 1) {
        entry.arrayIndex++;
        minHeap.add(entry);
      }
    }

    return res;
  }
}

File: lc/coreprep/implementation/patterns_reference_implementation/k_way_merge/merge_k_sorted_arrays.cpp
#include <algorithm>
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

struct Entry {
  int list_idx;
  int arr_idx;

  Entry(int list_idx, int arr_idx) : list_idx(list_idx), arr_idx(arr_idx) {}
};

struct CompareEntry {
  const vector<vector<int>> &a;

  CompareEntry(const vector<vector<int>> &a) : a(a) {}

  bool operator()(const Entry &e1, const Entry &e2) {
    return a[e1.list_idx][e1.arr_idx] > a[e2.list_idx][e2.arr_idx];
  }
};

// O(nlogk) / O(k)
vector<int> merge_k_sorted_arrays(const vector<vector<int>> &lists) {
  CompareEntry comp(lists);
  priority_queue<Entry, vector<Entry>, CompareEntry> min_heap(comp);

  for (int i = 0; i < lists.size(); i++) {
    min_heap.push(Entry(i, 0));
  }

  vector<int> res;
  while (!min_heap.empty()) {
    Entry entry = min_heap.top();
    min_heap.pop();

    res.push_back(lists[entry.list_idx][entry.arr_idx]);

    if (entry.arr_idx < lists[entry.list_idx].size() - 1) {
      entry.arr_idx++;
      min_heap.push(entry);
    }
  }

  return res;
}

int main() {
  ios::sync_with_stdio(0);
  cin.tie(0);

  int tt, k, n;
  cin >> tt;

  while (tt--) {
    cin >> k;

    vector<vector<int>> lists;
    for (int i = 0; i < k; i++) {
      cin >> n;

      vector<int> a(n);
      for (int j = 0; j < n; j++) {
        cin >> a[j];
      }
      lists.emplace_back(a);
    }

    auto merged = merge_k_sorted_arrays(lists);
    for (int m : merged) {
      cout << m << " ";
    }
    cout << "\n";
  }

  return 0;
}